Plugin tutorial, more changes (#4570)
haskell/haskell-language-server: docs/contributing/plugin-tutorial.md

## Anatomy of a plugin

HLS plugins are values of the `PluginDescriptor` datatype, which is defined in `hls-plugin-api/src/Ide/Types.hs` as:
```haskell
data PluginDescriptor (ideState :: Type) =
  PluginDescriptor { pluginId                   :: !PluginId
                   , pluginRules                :: !(Rules ())
                   , pluginCommands             :: ![PluginCommand ideState]
                   , pluginHandlers             :: PluginHandlers ideState
                   , pluginNotificationHandlers :: PluginNotificationHandlers ideState
                   , [...] -- Other fields omitted for brevity.
                   }
```
A plugin has a unique ID, command handlers, request handlers, notification handlers and rules:

* Request handlers respond to requests from an LSP client. They must fulfill these requests as quickly as possible.
* Notification handlers receive notifications from code not directly triggered by a user or client.
* Rules add new targets to the Shake build graph. Most plugins do not need to define new rules.

> * Commands are an LSP abstraction for user-initiated actions that the server handles. These actions can be long-running and involve multiple modules.

Author: This is a good summary of the LSP communication scheme. I moved part of the summary before "The anatomy of a plugin" into a subsection called "Notes about the Language Server Protocol" and tried to consolidate the rest. I think the text has improved a lot.

> * Commands are an LSP abstraction for user-initiated actions that the server handles. These actions can be long-running and involve multiple modules.

Reviewer: Thanks, I agree :)

## The explicit imports plugin

To achieve our plugin goals, we need to define:
- a command handler (`importLensCommand`),
- a code lens request handler (`lensProvider`).

> These will be assembled together in the `descriptor` function of the plugin, which contains all the information wrapped in the `PluginDescriptor` datatype mentioned above.

Reviewer: ```suggestion
These will be assembled in the `descriptor` function of the plugin, which contains all the information wrapped in the `PluginDescriptor` datatype mentioned above.
```

Using the convenience `defaultPluginDescriptor` function, we can bootstrap the plugin with the required parts:

```haskell
-- plugins/hls-explicit-imports-plugin/src/Ide/Plugin/ExplicitImports.hs

-- | The "main" function of a plugin.
descriptor :: Recorder (WithPriority Log) -> PluginId -> PluginDescriptor IdeState
descriptor recorder plId =
  (defaultPluginDescriptor plId)
    { pluginCommands = [importLensCommand], -- The plugin provides a command handler
      pluginHandlers = mconcat -- The plugin provides request handlers
        [ lensProvider
        ]
    }
```

We'll start with the command, since it's the simplest of the two.

### The command handler

> In short, commands work like this:

Reviewer: ```suggestion
In short, LSP commands work like this:
```

- The LSP server (HLS) initially sends a command descriptor to the client, in this case as part of a code lens.

> - Whenever the client decides to execute the command on behalf of a user (in this case a click on the code lens), it sends this same descriptor back to the LSP server. The server then handles and executes the command; this latter part is implemented by the `commandFunc` field of our `PluginCommand` value.

Reviewer: ```suggestion
- When the user clicks on the code lens, the client asks HLS to execute the command with the given descriptor. The server then handles and executes the command; this latter part is implemented by the `commandFunc` field of our `PluginCommand` value.
```

importLensCommand =
  PluginCommand
    { commandId = "ImportLensCommand"
    , commandDesc = "Explicit import command"
    , commandFunc = runImportCommand
    }

-- | Not implemented yet.
runImportCommand = undefined
```

The most important (and still `undefined`) field is `commandFunc :: CommandFunction`, a type synonym from `LSP.Types`:

```haskell
-- hls-plugin-api/src/Ide/Types.hs

type CommandFunction ideState a
  = ideState
  -> a
  -> LspM Config (Either ResponseError Value)
```

`CommandFunction` takes an `ideState` and a JSON-encodable argument.

Our handler will ignore the state argument and only use the `WorkspaceEdit` argument.

```haskell
-- | The type of the parameters accepted by our command
newtype ImportCommandParams = ImportCommandParams WorkspaceEdit
  deriving (Generic)
  deriving anyclass (FromJSON, ToJSON)

-- | The actual command handler
runImportCommand :: CommandFunction IdeState ImportCommandParams
runImportCommand _ (ImportCommandParams edit) = do
  -- This command simply triggers a workspace edit!
  _ <- sendRequest SWorkspaceApplyEdit (ApplyWorkspaceEditParams Nothing edit) (\_ -> pure ())
  return (Right Null)
```

> `runImportCommand` [sends a request](https://hackage.haskell.org/package/lsp/docs/Language-LSP-Server.html#v:sendRequest) to the client using the method `SWorkspaceApplyEdit` and the parameters `ApplyWorkspaceEditParams Nothing edit`, providing a response handler that does nothing. It then returns `Right Null`, which is an empty `Aeson.Value` wrapped in `Right`.

Reviewer: Maybe explain why we have an `Either` type here, and that `Left` is used for failures?

Author: I have moved the description of the Either up to when we introduce `CommandFunction`. Please have a look.

### The code lens provider

The code lens provider implements all the steps of the algorithm described earlier:

>  1. Request the type checking artifacts.
>  2. Extract the actual import lists from the type-checked AST.
>  3. Ask GHC to produce the minimal import lists for this AST.
>  4. For each import statement lacking an explicit list, determine its minimal import list and generate a code lens displaying this list along with a command to insert it.

> The provider takes the usual `LspFuncs` and `IdeState` arguments, as well as a `CodeLensParams` value containing a file URI. It returns an IO action that produces either an error or a list of code lenses for that file.

Reviewer: `LspFuncs` does not exist anymore, they're now part of the plugin monad.

Author: Here I do not know what to do. This seems like an integral part of the tutorial. Should we just leave it like that at the moment and add an issue to fix the tutorial later?

Reviewer 2: We are updating the tutorial right now, I don't think it makes sense to update the tutorial to still contain outdated info.

Reviewer 2: Also, there are exactly two mentions of `LspFuncs`, why is it integral? Let's just delete the reference here and in the `provider` function just below it. The type `CodeLensProvider` doesn't exist any more, so that needs to be updated anyway, right?

The command handler will be called `importLensCommand` and have the `PluginCommand` type, a type defined in `Ide.Types` as:

```haskell
-- hls-plugin-api/src/Ide/Types.hs

data PluginCommand ideState = forall a. (FromJSON a) =>
  PluginCommand { commandId   :: CommandId
                , commandDesc :: T.Text
                , commandFunc :: CommandFunction ideState a
                }
```

Let's start by creating an unfinished command handler. We'll give it an ID and a description for now:

```haskell
-- | The command handler.
importLensCommand :: PluginCommand IdeState
importLensCommand =
  PluginCommand
    { commandId = "ImportLensCommand"
    , commandDesc = "Explicit import command"
    , commandFunc = runImportCommand
    }

-- | Not implemented yet.
runImportCommand = undefined
```

The most important (and still `undefined`) field is `commandFunc :: CommandFunction`, a type synonym from `LSP.Types`:

```haskell
-- hls-plugin-api/src/Ide/Types.hs

type CommandFunction ideState a
  = ideState
  -> a
  -> LspM Config (Either ResponseError Value)
```

> `CommandFunction` takes an `ideState` and a JSON-encodable argument. `LspM` is a monad transformer with access to IO, and having access to a language context environment `Config`. The action evaluates to an `Either` value. `Left` indicates failure with a `ResponseError`, `Right` indicates sucess with a `Value`.

Author: Thank you!

## Wrapping up

There's only one Haskell code change left to do at this point: "link" the plugin in the `HlsPlugins` HLS module.

Integrating the plugin into HLS itself requires changes to several configuration files.

> A good approach is to use the ID of an existing plugin (e.g., `hls-class-plugin`) as a guide:

Reviewer: ```suggestion
A good approach is looking for the ID (f.e. `hls-class-plugin`) of an existing plugin as a reference:
```

Author: I have changed this to

```
+A good approach is to search for the ID of an existing plugin (e.g., `hls-class-plugin`):
```

Reviewer: That is much clearer, good call!
